### PR TITLE
Make stashing timer messages work 

### LINKED ISSUE #24557: stashed timer messages are discarded
I've recently bumped into a case that messages scheduled with timer and stashed via `stash()` will be discarded. It comes into play with any class that has `StashSupport` in the traits.
What happens is:
1. `Timers.aroundReceive` calls `super.aroundReceive` [here](https://github.com/akka/akka/blob/14bb878da17231bf0713bd636d2bc64aa0f04641/akka-actor/src/main/scala/akka/actor/Timers.scala#L40) with *unwrapped* message.
2. In case of a call to `stash`, the message is taken [here](https://github.com/akka/akka/blob/14bb878da17231bf0713bd636d2bc64aa0f04641/akka-actor/src/main/scala/akka/actor/Stash.scala#L157) and it's `actorCell.currentMessage`, so it's efficiently `Envelope` with `message` set to `TimerSchedulerImpl.TimerMsg` being stashed.
3. When unstashed the message gets discarded as it was already processed by `Timers`.

imho, supersedes #24076, because of being a more generic `Timers` and `StashSupport` problem.
somewhat related to #24080 as well.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #24557

Not 100% sure it is a good idea, but played around with other workarounds first but couldn't see a way to make them work given that Stash looks directly at `actorCell.currentMessage`.

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/Timers.scala [hunk 1]
```diff
@@ -37,8 +39,11 @@ trait Timers extends Actor {
     msg match {
       case timerMsg: TimerSchedulerImpl.TimerMsg ⇒
         _timers.interceptTimerMsg(timerMsg) match {
-          case OptionVal.Some(m) ⇒ super.aroundReceive(receive, m)
-          case OptionVal.None    ⇒ // discard
+          case OptionVal.Some(m) ⇒
+            // this is important for stash interaction, as stash will look directly at currentMessage #24557
+            actorCell.currentMessage = Envelope(m, sender())
+            super.aroundReceive(receive, m)
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: I guess we could do an instanceof-check on `this` and only do the extra work/allocation when `Stash` is mixed in?

File: akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala [hunk 2]
```diff
@@ -264,3 +264,41 @@ abstract class AbstractTimerSpec extends AkkaSpec {
...
+        unstashAll()
+      case "scheduled" ⇒
+        probe ! "saw-scheduled"
+        stash()
+    }
+  }
+
+  "Timers combined with stashing" should {
+
+    "something" in {
```
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala]: something

File: akka-actor/src/main/scala/akka/actor/Timers.scala [hunk 3]
```diff
@@ -37,8 +39,11 @@ trait Timers extends Actor {
     msg match {
       case timerMsg: TimerSchedulerImpl.TimerMsg ⇒
         _timers.interceptTimerMsg(timerMsg) match {
-          case OptionVal.Some(m) ⇒ super.aroundReceive(receive, m)
-          case OptionVal.None    ⇒ // discard
+          case OptionVal.Some(m) ⇒
+            // this is important for stash interaction, as stash will look directly at currentMessage #24557
+            actorCell.currentMessage = Envelope(m, sender())
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: um, nasty, but I can't find a better way

File: akka-actor/src/main/scala/akka/actor/Timers.scala (hunk 1 above)
Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/Timers.scala]: I think that would be good

**Comment by patriknw** [comment]
Somewhat related https://github.com/akka/akka/pull/24205

**Comment by 412b** [comment]
@johanandren, I've been checking this kind of solution as well and considered it being dirty, but the least effort :)
1. perhaps, `cell.currentMessage = cell.currentMessage.copy(message = m)`
2. in case of timers it seemed to me that it's not guaranteed to mixed exclusively with `ActorCell`. perhaps, smth like
`
if (context.isInstanceOf[ActorCell]) {
  val cell = context.asInstanceOf[ActorCell]
  cell.currentMessage = cell.currentMessage.copy(message = m)
}
super.aroundReceive(receive, m)
`
we have `Timers` in cluster sharding actors and due to 3kk of them every field is being count :)

File: akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala]: ups

**Comment by johanandren** [comment]
@412b there is never a case when context is not an `ActorCell` and `cell` is a method, so no extra field overhead.

**Comment by 412b** [comment]
@johanandren, yep, misread that `actorCell` is actually a `def` :smile:

**Review by 2m — APPROVED** [review]

**Review by ktoso — APPROVED** [review]
LGTM, nasty but let's do it

**Pull request merged**